File: rxgooglemap/src/main/java/com/sdoward/rxgooglemap/MarkerDragFunc.java
package com.sdoward.rxgooglemap;

import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.model.Marker;
import com.sdoward.rxgooglemap.events.*;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.functions.Function;

class MarkerDragFunc implements Function<GoogleMap, Observable<DragEvent>> {

    @Override
    public Observable<DragEvent> apply(final GoogleMap googleMap) {
        return Observable.create(new ObservableOnSubscribe<DragEvent>() {
            @Override
            public void subscribe(final ObservableEmitter<DragEvent> emitter) {
                GoogleMap.OnMarkerDragListener markerDragListener =
                        new GoogleMap.OnMarkerDragListener() {

                            @Override
                            public void onMarkerDragStart(Marker marker) {
                                emitter.onNext(new DragStartEvent(marker));
                            }

                            @Override
                            public void onMarkerDrag(Marker marker) {
                                emitter.onNext(new DragEvent(marker));
                            }

                            @Override
                            public void onMarkerDragEnd(Marker marker) {
                                emitter.onNext(new DragEndEvent(marker));
                            }
                        };
                googleMap.setOnMarkerDragListener(markerDragListener);
            }
        });
    }
}

File: rxgooglemap/src/main/java/com/sdoward/rxgooglemap/MarkerDragFunc.kt
package com.sdoward.rxgooglemap

import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.Marker
import com.sdoward.rxgooglemap.events.DragEndEvent
import com.sdoward.rxgooglemap.events.DragEvent
import com.sdoward.rxgooglemap.events.DragStartEvent
import io.reactivex.Observable
import io.reactivex.Observable.create
import io.reactivex.functions.Function


class MarkerDragFunc : Function<GoogleMap, Observable<DragEvent>> {

    override fun apply(t: GoogleMap): Observable<DragEvent> {
        return create { subscriber ->

            val markerDragListener: GoogleMap.OnMarkerDragListener = object : GoogleMap.OnMarkerDragListener {
                override fun onMarkerDragStart(marker: Marker) {
                    subscriber.onNext(DragStartEvent(marker))
                }

                override fun onMarkerDrag(marker: Marker) {
                    subscriber.onNext(DragStartEvent(marker))
                }

                override fun onMarkerDragEnd(marker: Marker) {
                    subscriber.onNext(DragEndEvent(marker))
                }

            }
            t.setOnMarkerDragListener(markerDragListener)

        }
    }

}